File: tests/runtimes/python/python.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typegraph import t, typegraph
from typegraph.graph.typegraph import Graph
from typegraph.policy import Policy
from typegraph.runtimes.python import PythonRuntime


def test(x):
    return x["a"]


def identity(x):
    return x["input"]


def stackoverflow(x):
    def fn():
        return fn()

    if x["enable"]:
        fn()
    return x["enable"]


def infinite_loop(x):
    import time

    # while x["enable"]:
    #     print("tic tac")
    # blocking
    time.sleep(20000)  # still blocking main thread
    return x["enable"]


tpe = t.struct({"a": t.integer(), "b": t.struct({"c": t.list(t.string())})})


@typegraph()
def python(g: Graph):
    public = Policy.public()
    python = PythonRuntime()

    g.expose(
        test=python.from_lambda(
            t.struct({"a": t.string()}),
            t.string(),
            lambda x: x["a"],
        ).with_policy(public),
        testDef=python.from_def(
            t.struct({"a": t.string()}),
            t.string(),
            test,
        ).with_policy(public),
        testMod=python.import_(
            t.struct({"name": t.string()}),
            t.string(),
            module="py/hello.py",
            deps=["py/nested/dep.py"],
            name="sayHello",
        ).with_policy(public),
        identity=python.from_def(
            t.struct({"input": tpe}),
            tpe,
            identity,
        ).with_policy(public),
        stackOverflow=python.from_def(
            t.struct({"enable": t.boolean()}),
            t.boolean(),
            stackoverflow,
        ).with_policy(public),
        infiniteLoop=python.from_def(
            t.struct({"enable": t.boolean()}),
            t.boolean(),
            infinite_loop,
        ).with_policy(public),
    )

File: tests/runtimes/python/python.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Policy, t, typegraph } from "@typegraph/sdk";
import { PythonRuntime } from "@typegraph/sdk/runtimes/python";
import outdent from "outdent";

const tpe = t.struct({
  a: t.string(),
  b: t.list(t.either([t.integer(), t.string()])),
});

export const tg = await typegraph("python", (g: any) => {
  const python = new PythonRuntime();
  const pub = Policy.public();

  g.expose({
    identityLambda: python
      .fromLambda(t.struct({ input: tpe }), tpe, {
        code: "lambda x: x['input']",
      })
      .withPolicy(pub),
    identityDef: python
      .fromDef(t.struct({ input: tpe }), tpe, {
        code: outdent`
        def identity(x):
          return x['input']
        `,
      })
      .withPolicy(pub),
    identityMod: python
      .import(t.struct({ input: tpe }), tpe, {
        name: "identity",
        module: "py/hello.py",
        deps: ["py/nested/dep.py"],
      })
      .withPolicy(pub),
  });
});
